// @line/bot-sdk/lib/exceptions.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var SignatureValidationFailed = (function (_super) {
    __extends(SignatureValidationFailed, _super);
    function SignatureValidationFailed(message, signature) {
        _super.call(this, message);
        this.signature = signature;
    }
    return SignatureValidationFailed;
}(Error));
exports.SignatureValidationFailed = SignatureValidationFailed;
var JSONParseError = (function (_super) {
    __extends(JSONParseError, _super);
    function JSONParseError(message, raw) {
        _super.call(this, message);
        this.raw = raw;
    }
    return JSONParseError;
}(Error));
exports.JSONParseError = JSONParseError;
var RequestError = (function (_super) {
    __extends(RequestError, _super);
    function RequestError(message, code, originalError) {
        _super.call(this, message);
        this.code = code;
        this.originalError = originalError;
    }
    return RequestError;
}(Error));
exports.RequestError = RequestError;
var ReadError = (function (_super) {
    __extends(ReadError, _super);
    function ReadError(originalError) {
        _super.call(this, originalError.message);
        this.originalError = originalError;
    }
    return ReadError;
}(Error));
exports.ReadError = ReadError;
var HTTPError = (function (_super) {
    __extends(HTTPError, _super);
    function HTTPError(message, statusCode, statusMessage, originalError) {
        _super.call(this, message);
        this.statusCode = statusCode;
        this.statusMessage = statusMessage;
        this.originalError = originalError;
    }
    return HTTPError;
}(Error));
exports.HTTPError = HTTPError;
