// @line/bot-sdk/lib/urls.js
"use strict";
var qs = require("querystring");
var baseURL = process.env.API_BASE_URL || "https://api.line.me/v2/bot/";
var apiURL = function (path, query) {
    return baseURL + path + (query ? "?" + qs.stringify(query) : "");
};
exports.reply = apiURL("message/reply");
exports.push = apiURL("message/push");
exports.multicast = apiURL("message/multicast");
exports.content = function (messageId) {
    return apiURL("message/" + messageId + "/content");
};
exports.profile = function (userId) { return apiURL("profile/" + userId); };
exports.groupMemberProfile = function (groupId, userId) {
    return apiURL("group/" + groupId + "/member/" + userId);
};
exports.roomMemberProfile = function (roomId, userId) {
    return apiURL("room/" + roomId + "/member/" + userId);
};
exports.groupMemberIds = function (groupId, start) {
    return apiURL("group/" + groupId + "/members/ids", start ? { start: start } : null);
};
exports.roomMemberIds = function (roomId, start) {
    return apiURL("room/" + roomId + "/members/ids", start ? { start: start } : null);
};
exports.leaveGroup = function (groupId) { return apiURL("group/" + groupId + "/leave"); };
exports.leaveRoom = function (roomId) { return apiURL("room/" + roomId + "/leave"); };
exports.richMenu = function (richMenuId) {
    return apiURL("richmenu" + (richMenuId ? "/" + richMenuId : ""));
};
exports.richMenuList = function () { return apiURL("richmenu/list"); };
exports.userRichMenu = function (userId, richMenuId) {
    return apiURL(("user/" + userId + "/richmenu") + (richMenuId ? "/" + richMenuId : ""));
};
exports.richMenuContent = function (richMenuId) {
    return apiURL("richmenu/" + richMenuId + "/content");
};
